### DISCUSSION TITLE
Best way to serialize localStorage to prefs? Load workspace by URL?

### DISCUSSION THREAD

**Discussion by echuber2** [eclipse-theia/theia#8308] [General] [answered]
Hi, I think this has been discussed a bit in the past here and there, but I'm still not sure what the best approach should be as of today:

I'm trying to run Theia Docker instances with a workspace mounted as a volume, where the volume can persist even when the container spins down. So far, this seems to work okay; you can resume your work when you reload the Docker. However, Theia seems to store many details of the workspace (open directory, open panes, layout) in the localStorage, not the preferences file. This means the workspace state is tied to the browser session rather than persisted in the volume.

Right now, if you resume working on another computer, when you load up there is no active workspace. You can select the recent workspace to load it then, but it still loads without any reasonable layout (no terminal open, no file browser pane open, etc.). I'd basically like some way to make the default presentation on launch resemble VS Code a bit more.

Here are the two main things I'd like to do somehow:

- I would like some way to serialize and deserialize the localStorage to a preferences file so that layout and workspace can persist, if possible.
- Apart from that, I wish the URL could specify the workspace and folder to open on a fresh launch (with empty localStorage). So the workspace would open with at least the correct folder displayed and some sane default panes open (terminal, file browser, etc.).

Is there an extension that does this easily? Is anyone working on one? Thanks.

**Comment by akosyakov** [comment, marked as answer]
User preferences by definition workspace agnostic and putting layout information of concrete user in workspace preference is not practical. Depending on the product one can customize `ShellLayoutRestorer` or `StorageService`. For instance Gitpod is using database to store them. You will need to look into custom Theia extension to support something like that in Docker context.

**Reply by echuber2** [reply]
I can look into doing this. Are there any specific examples? My use case is embedding Theia for educational purposes. When a student spins up an instance of the container, I'd like to load a prepared workspace with a specific file loaded, the terminal pane open, etc. (Instead, Theia loads with no workspace and no panes open, although I can direct people on what steps to do next.)

**Reply by akosyakov** [reply]
For that purpose, you need to implement `FrontendApplicationContribution.initializeLayout` lifecycle hook. It is called when someone opens a project for the first time without any layout information. Look in this repo how it is implemented. If you want to opea a specific fie there is `OpenerService.getOpener.open` to which you pass an URI of the file. There is also `TerminalFrontendContribution` which allow to create and reveal a new terminal. You can call them from the hook above.

**Reply by echuber2** [reply]
Okay, thank you. I'll study this for a while.

**Reply by echuber2** [reply]
Would it be inadvisable to let initial layout information be stored in a dotfile within a workspace, if not in the preferences? I could author a plugin to read that information, if found, and use it to initialize the workspace based on the particular expectations of each assignment.

**Reply by akosyakov** [reply]
We use DI as a primary means of customisations. 3rd party extension could allow something like that.

> Would it be inadvisable to let initial layout information be stored in a dotfile within a workspace, if not in the preferences? 

So yes, you can do it for your case.
